### ISSUE TITLE
batch from DataLoader with SSDDefaultTrainTransform(width, height, anchors) seems error

### ISSUE THREAD

**Issue by wyattsuen** [dmlc/gluon-cv#906]
I followed the example https://gluon-cv.mxnet.io/build/examples_detection/train_ssd_voc.html to train my own dataset with model ssd_300_vgg16_atrous_custom with only 1 class.

This issue I counted recently like the old closed issue https://github.com/dmlc/gluon-cv/issues/470, but I find no answer.

When I use SSDDefaultTrainTransform(width, height) without anchors, I checked the batch[0] and batch[1], all is right.
```
train_dataset=LstDetection('./train.lst', root=root_path)
train_transform=presets.ssd.SSDDefaultTrainTransform(width, height)
batchify_fn=Tuple(Stack(), Pad(pad_val=-1))
train_loader=gdata.DataLoader(train_dataset.transform(train_transform), 
            batch_size=1, 
            shuffle=False, 
            batchify_fn=batchify_fn, 
            last_batch='rollover', 
            num_workers=0)
for batch in train_loader:
    batch[0], batch[1]
```

But when I use SSDDefaultTrainTransform(width, height, anchors), the batch[1], batch[2] seems wrong.
```
train_transform=presets.ssd.SSDDefaultTrainTransform(width, height, anchors)
batchify_fn=Tuple(Stack(), Stack(), Stack())
train_loader=gdata.DataLoader(
    train_dataset.transform(train_transform), 
    batch_size=1, 
    shuffle=False, 
    batchify_fn=batchify_fn, 
    last_batch='rollover', 
    num_workers=num_workers
)
for batch in train_loader:
    batch[0], batch[1], batch[2]
```

Following is my own debug:
```
with autograd.train_mode():
    cls_preds, box_preds, anchors=net(x)
```
The anchors.shape is (1,8732,4),  and the content seems right.
![image](https://user-images.githubusercontent.com/42318414/62940114-6b8c9500-be05-11e9-99ba-87e0d0c6b460.png)

batch[1].shape is (1, 8732), most member of it is 0. I find out the index of 1 using:
```
cls_pos=np.argwhere(batch[1][0].asnumpy()==1)
```
![image](https://user-images.githubusercontent.com/42318414/62940413-1d2bc600-be06-11e9-8a07-4ebe294cceb6.png)

I think it means the bounding box of index some number like above can all fit my dataset with only 1 class.

But I need the target bounding box corners. That's in batch[2]. So I check the batch[2].  Its shape is (1,8732,4). The batch[2] most memebers seem to be [0. 0. 0. 0.]. I think it means no category. I find out the bouding box with category==1 using:
```
batch[2][0][cls_pos]
```
![image](https://user-images.githubusercontent.com/42318414/62941119-e060ce80-be07-11e9-97a5-f2781afc391a.png)
the bounding box [xmin, ymin, xmax, ymax] seems from "-2 to 2" value，but not the absolute coordinate! I think gluoncv use absolute coordinate of bouding box and mxnet use the relative coordinate! What's the "-2 to 2" value means?

I want to check the image and the first bounding box.
```
img=batch[0][0]
img=img.transpose((1, 2, 0)) * nd.array((0.229, 0.224, 0.225)) + nd.array((0.485, 0.456, 0.406))
img=(img*255).clip(0,255)
cls_np=batch[1][0].asnumpy()
cls_pos=np.argwhere(cls_np==1)
cls=batch[1][0][cls_pos]
bbox=batch[2][0][cls_pos]
print(bbox)
ax = utils.viz.plot_bbox(img.asnumpy(), bbox[0], labels=cls[0])
plt.show()
```
![image](https://user-images.githubusercontent.com/42318414/62941739-5b76b480-be09-11e9-9f12-282dd00b58f1.png)
The first target bounding box is on the left top corner. The red rectangle is my own hand real label bounding box. What's wrong?

**Comment by zhreshold** [comment]
you are right that batch[2] is training targets, however, they are encoded values computed by https://github.com/dmlc/gluon-cv/blob/059e6deb060a7b0368a09d43614558e4c110f4bc/gluoncv/nn/coder.py#L128-L131 
That's why the values are from -2 to 2. To get real ground truth bounding boxes, you can revert the values by using BBoxDecoder or retrive the values in GrountTruth.

**Comment by wyattsuen** [comment]
@zhreshold，thanks very much for the answer. But it can't solve my problem. If the batch[2] in training mode use normalized value, as following, I believe the box_pred will also be normalized value.
```
mbox_loss=SSDMultiBoxLoss()
with autograd.record():
            cls_pred, box_pred, anchors=net(batch[0])
            sum_loss, cls_loss, box_loss=mbox_loss(cls_pred, box_pred, batch[1], batch[2])
```
But in the predict mode, I follow the gluoncv tutorial https://gluon-cv.mxnet.io/build/examples_detection/demo_ssd.html, using following code:
```
class_IDs, scores, bounding_boxes = net(x)
ax = utils.viz.plot_bbox(img, bounding_boxes[0], scores[0],class_IDs[0], class_names=net.classes)
plt.show()
```
The tutorial doesn't use the BBoxDecoder to decode the bounding_boxes, so the bounding_boxes in predict mode must be absolute coordinate. But in my own trained finished model, the bounding_boxes in predict mode are far far away from my expect. I think the bounding box coordinate is in chaos for gluoncv, which causes some troubles. 
So should I decode the bounding box coordinate in training mode or in predict mode?

**Comment by zhreshold** [comment]
If anchor is not provided to SSDDefaultTransform, it will return raw bounding boxes, otherwise it will return encoded training targets.

https://github.com/dmlc/gluon-cv/blob/123af8cf9f15a879c16a5c7d12f01ce1471d85b6/gluoncv/data/transforms/presets/ssd.py#L177-L185

Can you explain what's your original problem? I have no idea what's the problem so far.

**Comment by wyattsuen** [comment]
@zhreshold I trained the 'ssd_300_vgg16_atrous_custom' model with only 1 class and my own pictures following the tutorial. And when I predict the test pictures, the predicted bounding box is far far away from my expected bounding box. I have to debug the bounding box coordinate of input and output. But after your answer, I think the bouding box coordinate is no wrong. So where is wrong?

Now I have some new idea about my problem. If I use SSDDefaultTrainTransform(width, height) without anchors or use SSDDefaultTrainTransform(width, height, anchors), the following predict bounding_boxes coordinate may be different?
```
class_IDs, scores, bounding_boxes = net(x)
```
As you noted, I think maybe the predicted bounding_boxes need to decode after SSDDefaultTrainTransform(width, height, anchors) and need not to decode after SSDDefaultTrainTransform(width, height). Is it right?

So I prepare to test to decode the bounding_boxes in predict result after SSDDefaultTrainTransform(width, height, anchors). But when I search the BBoxDecoder class, I even can't find the class in gluoncv.nn.coder. And I even can't find this class in all gluoncv document.  Would you kindly pls. give some example how to decode the bounding box coordinate?
